Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -524,60 +532,68 @@ public final class ManifestLoader: ManifestLoaderProtocol {
-            fileSystem: fileSystem
-        )
+        let key : CacheKey
+        do {
+            key = try CacheKey(
+                packageIdentity: packageIdentity,
+                manifestPath: path,
+                toolsVersion: toolsVersion,
+                env: ProcessEnv.vars,
+                swiftpmVersion: SwiftVersion.currentVersion.displayString,
+                fileSystem: fileSystem
+            )
+        } catch {
+            return completion(.failure(error))
+        }
 
         do {
             // try to get it from the cache
             if let result = try cache?.get(key: key.sha256Checksum), let manifestJSON = result.manifestJSON, !manifestJSON.isEmpty {
-                return try self.parseManifest(
+                return completion(.success(try self.parseManifest(
                     result,
                     packageIdentity: packageIdentity,
                     packageKind: packageKind,
                     toolsVersion: toolsVersion,
                     identityResolver: identityResolver,
                     fileSystem: fileSystem,
                     observabilityScope: observabilityScope
-                )
+                )))
             }
         } catch {
             observabilityScope.emit(warning: "failed loading cached manifest for '\(key.packageIdentity)': \(error)")
+            return completion(.failure(error))
         }
 
         // shells out and compiles the manifest, finally output a JSON
-        let result = self.evaluateManifest(
+        self.evaluateManifest(
             packageIdentity: key.packageIdentity,
             manifestPath: key.manifestPath,
             manifestContents: key.manifestContents,
             toolsVersion: key.toolsVersion,
             delegateQueue: delegateQueue
-        )
+        ) { result in
+            do {
+                let evaluationResult = try result.get()
+                // only cache successfully parsed manifests
+                let parseManifest = try self.parseManifest(
+                    evaluationResult,
+                    packageIdentity: packageIdentity,
+                    packageKind: packageKind,
+                    toolsVersion: toolsVersion,
+                    identityResolver: identityResolver,
+                    fileSystem: fileSystem,
+                    observabilityScope: observabilityScope
+                )
 
-        // only cache successfully parsed manifests
-        let parseManifest = try self.parseManifest(
-            result,
-            packageIdentity: packageIdentity,
-            packageKind: packageKind,
-            toolsVersion: toolsVersion,
-            identityResolver: identityResolver,
-            fileSystem: fileSystem,
-            observabilityScope: observabilityScope
-        )
+                // FIXME: (diagnostics) pass in observability scope when we have one
+                try cache?.put(key: key.sha256Checksum, value: evaluationResult)
 
-        do {
```

Comment:
here we actually want two different do blocks since we emit a warning that is specific to failure to store the manifest in the cache

Reply:
@pcbeard this still needs to be split into two "do" blocks? otherwise errors from parseManifest will emit "failed storing manifest..."

Reply:
This was addressed in 4791c658.